source/nodejs/adaptivecards-designer/src/components/treeitem.ts

```diff
@@ -0,0 +1,95 @@
+import { DesignerPeer } from "../card-designer";
+import * as Utils from "../utils";
+
+export default class TreeItem {
+    readonly owner: DesignerPeer;
+    private _listElement: HTMLElement;
+    private _childContainerElement: HTMLElement;
+    private _isExpanded: boolean = true;
+
+    constructor(owner: DesignerPeer) {
+        this.owner = owner;
+    }
+
+    render(indentationLevel: number = 0): HTMLElement {
+        let rootElement = document.createElement("div");
+        rootElement.className = "treeview__container";
+
+        this._listElement = this.internalRender(indentationLevel);
+        rootElement.appendChild(this._listElement);
+
+        if (this.owner.getChildCount() > 0) {
+            rootElement.appendChild(this.renderChildList(indentationLevel));
+        }
+
+        this.updateLayout();
+        return rootElement;
+    }
+
+    private renderChildList(indentationLevel: number): HTMLElement {
+        this._childContainerElement = document.createElement("div");
+        this._childContainerElement.className = "treeview__container";
+
+        for (let i = 0; i < this.owner.getChildCount(); i++) {
+            let currentItem = this.owner.getChildAt(i);
+            this._childContainerElement.appendChild(currentItem.treeItem.render(indentationLevel + 1));
+        }
+
+        if (!this._isExpanded) {
+            this._childContainerElement.classList.add("is-folded");
+        }
+
+        return this._childContainerElement;
+    }
+
+    private internalRender (indentationLevel: number): HTMLElement {
+        let listItem = document.createElement("div");
+        listItem.className = "treeview__element";
+
+        listItem.onclick = (e: MouseEvent) => {
+            this.owner.isSelected = true;
+
+            e.cancelBubble = true;
+            e.preventDefault();
+        }
+
+        if (this.owner.getChildCount() > 0) {
+            let foldArrow = document.createElement("button");
+            foldArrow.onclick = (e: MouseEvent) => {
+                e.cancelBubble = true;
+                e.preventDefault();
+                foldArrow.classList.toggle("is-rotated");
+                this.foldTreeViewContainer();
+            }
+
+            foldArrow.className = `treeview__icon treeview__icon--arrow ${!this._isExpanded ? "is-rotated" : ""}`;
+            listItem.appendChild(foldArrow);
+        }
+
+        let icon = document.createElement("span");
+        icon.className = `treeview__icon treeview__icon--${Utils.sanitizeString(this.owner.getCardObjectTypeName())}`;
```

Comment:
As I tried to explain in my previous review, the CSS class name (I called it iconUrl at the time) should be a property of IDesignerPeerRegistration. You don't want to dynamically build this class name like you do here, the main reason being that you have to be doing it here and in another place for the toolbox. That's one too many places.

Reply:
We don't understand what you mean. The thing is, that we add icons via CSS because they add differences between the .treeview__icon and the .aside-item__icon (positioning ones).

- Aside: toolbox panel
- Treeview: treeview panel

Reply:
I am saying that you are building a class name string using the result of getCardObjectTypeName appended to some prefix, and that has to match what you have in CSS. You are using the same icon classes to build the toolbox, and thus you most likely have a very similar line of code somewhere else. Building a class name like this isn't very solid overall IMO. A better approach would be to have a method, e.g. getIconCssClassName, that would be called from anywhere the icon is being used (tree and toolbox.) That method though can't be on DesignerPeer because you don't have DesignerPeer instances when you build the toolbox. So it needs to be somewhere else, and the right place is a property on IDesignerPeerRegistration.

Let's leave as is for now. I will take care of this once we are all done with this PR.